notes/inheritance-1.md

```diff
@@ -0,0 +1,329 @@
+
+This is why, in the previous chapter, we could do something like this:
+
+```C++
+void printRank(Animal &ani){
+    std::cout << ani.Rank() << std::endl;
+}
+
+int main(){
+    Animal a;
+    Bun b;
+    KingBun kb;
+    printRank(a); // prints 1
+    printRank(b); // prints 5
+    printRank(kb); // prints 10
+}
+```
+- Since `Bun` and `KingBun` are all derived classes of `Animal`, we can bind a reference (to `Animal`) to an object of type `Bun` or `KingBun`
+
+<br>
+
+Secondly, since we can do this type of binding, when we use a reference or pointer to a base-class type, we don't know the actual type of the object to which the pointer/reference is bound!
+- The object can be an object of the base class or one of its derived classes
+- In the example above, is `ani` bound to an `Animal`, `Bun`, or `KingBun` object? We don't know until runtime!
+
+---
+## Static type vs. Dynamic type
+
+The **static type** of an expression is always known at compile time
+- It is the type with which a variable is declared or that an expression yields
+
+For example, in our `printRank` function, parameter `ani` has static type `Animal&`
+
+<br>
+
+The **dynamic type** is the type of the object in memory that the variable or expression actually represents
+
+The dynamic type of an variable that isn't a reference or pointer is just the same as the variable's static type
+
+Example:
+```C++
+Bun b;
+```
+- Variable `b`'s static type is `Bun`, its dynamic type is also `Bun`
+- Another way to put it: a variable of type `Bun` is always a `Bun` object
+
+> However, for a pointer or reference, the dynamic type depends on the type of argument to which the pointer/reference is bound
+
+Example:
+```C++
+void printRank(Animal &ani){
+    // ani's static type is Animal&
+    std::cout << ani.Rank() << std::endl;
+}
+
+int main(){
+    Animal a;
+    Bun b;
+    KingBun kb;
+    printRank(a); // ani's dynamic type is Animal
+    printRank(b); // ani's dynamic type is Bun
+    printRank(kb); // ani's dynamic type is KingBun
+}
+
+```
+
+---
+## Base to derived conversion?
+
+You might be wondering: in C++, are we allowed to convert from a base class object to a derived class object?
+
+Technically yes, you can try using virtual destructors and `std::dynamic_cast` to perform a safe conversion
+
+Generally though, this isn't safe. Think back to our section on [subobjects](#derived-to-base-conversions). Most of the time, there is no guarantee that a base class object contains the members defined by the derived class.
+- Conceptual example: Say you needed the help of [Korosensei](https://ansatsukyoshitsu.fandom.com/wiki/Korosensei), an excellent teacher who can also fly at top speeds of Mach 20. Since I can't guarantee that any random teacher can also fly, I'm not allowed to perform a conversion
+
+Thus, there is no implicit conversion from base to derived
+```C++
+Animal a;
+Bun* bunPtr = &a;  // compile error, cannot convert base to derived
+Bun& bunRef = a; // compile error
+
```

Comment:
I believe using `std::dynamic_cast` is safe. The reason being that it will give you the casted object if it is an instance of the template object, and null otherwise! I might be wrong, but you're welcome to read more about it. You can then treat it as an `instanceOf()` operator just like in Java, and use it like you would in Java.

Reply:
Let me change this, I didn't make it very clear that I meant to talk about implicit conversions, not user-defined ones.